### PR TITLE
Implementation of CMA support

### PR DESCRIPTION
**Pull request by arporter**
This pull request completes issue #6.

### CODE REVIEW

**Comment by arporter** [comment]
Eike has identified one further extension to the functionality added in this pull request. "Matrix-matrix" kernels will be permitted to have zero or more read-only scalar arguments. I'll add this change in now so please hold off reviewing...

**Comment by arporter** [comment]
OK. I think I've satisfied Eike's new requirement. Over to Rupert for review...

File: doc/dynamo0p3.rst [hunk 1]
```diff
@@ -891,22 +894,34 @@ as the number of dofs for each of the dofmaps. The full set of rules is:
...
+	     degrees of freedom for the function space.
+
+	3) If the CMA operator has this space as its to/from space then
+	   include the column-banded dofmap, the list of offsets for the
+	   to/from-space. This is an integer array of rank 2. The first
+	   dimension is ``"ndf_"<arg_function_space>```` and the second
+           is ``nlayers``.
+
 
 Application/Inverse-Application
```
Comment (arporter) [inline doc/dynamo0p3.rst]: I think it is correct as this rule is for a kernel that applies a CMA operator which, by definition, has been assembled for a whole column - therefore we don't need explicit information on levels. The dofs we provide are (presumably) for the whole column. Does that make sense?

File: doc/dynamo0p3.rst [hunk 2]
```diff
@@ -966,11 +981,13 @@ and ``ncell_3d`` scalar arguments. The full set of rules are then:
        intent ``in``.
     2) Include the number of cells in the 2D mesh, ``ncell_2d``, which is
        an integer with intent ``in``.
-    3) For each (CMA operator) argument specifed in meta-data:
+    3) For each CMA operator argument specifed in meta-data:
 
        1) Include it and its associated parameters (see Rule 5 of CMA
 	  Assembly kernels).
 
+    4) For each scalar argument specified in meta-data include the
```
Comment (arporter) [inline doc/dynamo0p3.rst]: Well spotted. The code is right, the documentation is wrong - I've corrected it (bc5376c).

File: examples/check_examples [hunk 3]
```diff
@@ -42,6 +42,11 @@ python ../../../src/generator.py -s ./omp_script.py alg.x90
 python ../../../src/generator.py -s ./omp_reprod_script.py alg.x90
 cd -
 
+echo "Running example dynamo/eg7"
+cd dynamo/eg7
+python ../../../src/generator.py alg.x90
```
Comment (arporter) [inline examples/check_examples]: Good point. Feels good to do that :-) (70c93b1)

File: examples/dynamo/README [hunk 4]
```diff
@@ -7,9 +7,12 @@
...
+containing this README use the Dynamo 0.1 API. Those in eg3 - eg7 use
 version 0.3 of the Dynamo API. They are primarily provided to
-illustrate the use of the PSyclone code-generation system.
+illustrate the use of the PSyclone code-generation system. No guarantee
+is made as to their functional correctness or usefulness (i.e. the
+calculations that they perform may often be nonsensical - it is the use
+of PSyclone that is being illustrated).
 
 In order to use PSyclone you must first add it to your PYTHONPATH. The
 simplest way to do this is to change to the top-level directory of your
```
Comment (arporter) [inline examples/dynamo/README]: Done.

File: examples/dynamo/README [hunk 5]
```diff
@@ -57,22 +60,22 @@ algorithm layer.
...
     python ../../../src/generator.py alg.x90
 
-with OpenMP (and loop fusion)
+code parallelised with OpenMP (and loop fusion)
 ::
     cd eg6/
     python ../../../src/generator.py -nodm -s ./omp_script.py alg.x90
 
-or with both MPI and OpenMP (and loop fusion)
+or code parallelised with both MPI and OpenMP (and loop fusion)
```
Comment (arporter) [inline examples/dynamo/README]: Done both (e8e65fd)

File: src/dynamo0p3.py [hunk 6]
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
...
+                        "only LMA operator arguments".format(self.name))
+            else:
+                # A valid matrix-matrix kernel must only have CMA operators
+                # and scalars as arguments.
+                scalar_args = psyGen.args_filter(
                     self._arg_descriptors,
-                    arg_types=["gh_operator", "gh_field", "gh_scalar"],
-                    arg_accesses=["gh_inc", "gh_write"])
-                if write_args:
+                    arg_types=["gh_real", "gh_integer"])
```
Comment (arporter) [inline src/dynamo0p3.py]: Good idea - done.

File: src/dynamo0p3.py [hunk 7]
```diff
@@ -873,6 +875,13 @@ def func_descriptors(self):
         DynFuncDescriptor03 objects, one for each function space. '''
         return self._func_descriptors
 
+    @property
+    def cma_operation(self):
+        ''' Returns the type of CMA operation identified from the kernel
+        meta-data (one of 'assembly', 'apply' or 'matrix-matrix') or
+        "" if the kernel does not involve CMA operators '''
```
Comment (arporter) [inline src/dynamo0p3.py]: So it does. Fixed.

File: src/dynamo0p3.py [hunk 8]
```diff
@@ -1188,6 +1200,17 @@ def __init__(self, schedule):
         # We create a dictionary whose keys are the map names and entries
         # are the corresponding field objects.
         self._unique_fs_maps = {}
+        # We also create a dictionary of column-banded dofmaps. Entries
+        # in this one are a list where the first item is the
+        # corresponding CMA object and the second is whether the map
+        # corresponds to the 'to' or 'from' space.
+        self._unique_cbanded_maps = {}
```
Comment (arporter) [inline src/dynamo0p3.py]: I agree. I felt slightly uncomfortable when I did that but didn't pause to examine why. I like your suggestion so have changed it. 475e151.

File: src/dynamo0p3.py [hunk 9]
```diff
@@ -1200,6 +1223,40 @@ def __init__(self, schedule):
...
                         if map_name not in self._unique_fs_maps:
                             self._unique_fs_maps[map_name] = fld_arg
+                if call.cma_operation == "assembly":
+                    # A kernel that assembles a CMA operator requires
+                    # column-banded dofmaps for its 'to' and 'from'
+                    # function spaces
+                    cma_args = psyGen.args_filter(
+                        call.arguments.args,
+                        arg_types=["gh_columnwise_operator"])
+                    map_name = get_cbanded_map_name(
```
Comment (arporter) [inline src/dynamo0p3.py]: It wasn't too bad. Done this and the next one in 33a0c9e

File: src/dynamo0p3.py [hunk 10]
```diff
@@ -1209,33 +1266,174 @@ def initialise_dofmaps(self, parent):
...
+
+        # CMA operator indirection dofmaps
+        decl_ind_map_names = \
+            [dmap+"(:) => null()" for dmap in self._unique_indirection_maps]
+        if decl_ind_map_names:
+            parent.add(DeclGen(parent, datatype="integer", pointer=True,
+                               entity_decls=decl_ind_map_names))
+
+
+class DynInvokeCMAOperators(object):
```
Comment (arporter) [inline src/dynamo0p3.py]: I think it is better the way it is - the new dofmap entities fit well within the existing `DynInvokeDofmaps` class and they are conceptually different to the CMA operators themselves.

File: src/dynamo0p3.py [hunk 11]
```diff
@@ -1209,33 +1266,174 @@ def initialise_dofmaps(self, parent):
...
+    ''' Holds all information on the CMA operators required by an invoke '''
+
+    cma_same_fs_params = ["nrow", "bandwidth", "alpha",
+                          "beta", "gamma_m", "gamma_p"]
+    cma_diff_fs_params = ["nrow", "ncol", "bandwidth", "alpha",
```
Comment (arporter) [inline src/dynamo0p3.py]: This is the order that the parameters are passed in to the kernels so it matters. I've added a couple of comments.

File: src/dynamo0p3.py [hunk 12]
```diff
@@ -1569,17 +1794,28 @@ def gen_code(self, parent):
                 TypeDeclGen(invoke_sub,
                             datatype="operator_proxy_type",
                             entity_decls=op_proxy_decs))
+        cma_op_proxy_decs = self.unique_proxy_declarations(
+            "gh_columnwise_operator")
+        if len(cma_op_proxy_decs) > 0:
```
Comment (arporter) [inline src/dynamo0p3.py]: I'm not sure that was deliberate and your suggestion is more pythonic so I've changed my code (and resisted the temptation to change the others).

File: src/dynamo0p3.py [hunk 13]
```diff
@@ -1569,17 +1794,28 @@ def gen_code(self, parent):
...
+                TypeDeclGen(invoke_sub,
+                            datatype="columnwise_operator_proxy_type",
+                            entity_decls=cma_op_proxy_decs))
+
         # Initialise the number of layers
         invoke_sub.add(CommentGen(invoke_sub, ""))
         invoke_sub.add(CommentGen(invoke_sub, " Initialise number of layers"))
         invoke_sub.add(CommentGen(invoke_sub, ""))
 
         # Use the first argument that is not a scalar
```
Comment (arporter) [inline src/dynamo0p3.py]: Done.

File: src/dynamo0p3.py [hunk 14]
```diff
@@ -1569,17 +1794,28 @@ def gen_code(self, parent):
...
         first_var = None
+        cma_op = None
         for var in self.psy_unique_vars:
-            if var.type in ["gh_field", "gh_operator"]:
+            if not first_var and var.type in ["gh_field", "gh_operator",
```
Comment (arporter) [inline src/dynamo0p3.py]: Good idea - done.

File: src/dynamo0p3.py [hunk 15]
```diff
@@ -2345,6 +2605,13 @@ def _setup(self, ktype, module_name, args, parent):
                 label=self._qr_text)
 
     @property
+    def cma_operation(self):
+        ''' Returns the type of CMA operation performed by this kernel
+        (one of 'assembly', 'apply' or 'matrix-matrix') or None if the
```
Comment (arporter) [inline src/dynamo0p3.py]: Good spot. It returns None now (if kernel is not CMA-related).

File: src/dynamo0p3.py [hunk 16]
```diff
@@ -2619,11 +2886,24 @@ def generate(self):
...
             self.cell_position()
-        # always pass the number of layers in the mesh
-        self.mesh_height()
+        # Pass the number of layers in the mesh unless this kernel is
+        # applying a CMA operator or doing a CMA matrix-matrix calculation
+        if self._kern.cma_operation not in ["apply", "matrix-matrix"]:
+            self.mesh_height()
+        # Pass the number of cells in the mesh if this kernel has a
+        # LMA operator argument
+        # TODO this code should replace the code that currently includes
```
Comment (arporter) [inline src/dynamo0p3.py]: Another issue - we have a ticket (551) for it on SRS that I will migrate over to github.

File: src/dynamo0p3.py [hunk 17]
```diff
@@ -2849,9 +3193,11 @@ def scalar(self, scalar_arg):
     def fs_compulsory(self, function_space):
         '''add compulsory arguments common to operators and
         fields on a space.'''
-        # There is currently one compulsory argument: "ndf".
-        ndf_name = get_fs_ndf_name(function_space)
-        self._arglist.append(ndf_name)
+        if self._kern.cma_operation not in ["matrix-matrix"]:
+            # There is currently one compulsory argument: "ndf" but only
```
Comment (arporter) [inline src/dynamo0p3.py]: Good point :-) I've renamed it `fs_common` and done this for the base class too - 487524a.

File: src/dynamo0p3.py [hunk 18]
```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
...
+        ''' add the CMA operator arguments to the argument list '''
+        from f2pygen import DeclGen
+        # The CMA operator itself
+        self._arglist.append(arg.name)
+        # Associated scalar parameters
+        nrow = arg.name + "_nrow"
+        _local_args = [nrow]
+        if arg.function_space_to.orig_name != \
+           arg.function_space_from.orig_name:
+            # If the to- and from-spaces are the same then so are ncol and
```
Comment (arporter) [inline src/dynamo0p3.py]: I'm not quite sure what you mean. However, I've re-written the comment so that it is worded correctly for its location. By default we pass both `nrow` and `ncol` to a kernel but if the to- and from-spaces are the same then we only pass `nrow`.

File: src/dynamo0p3.py [hunk 19]
```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
...
+        ''' Declare the indirection dofmaps required when applying a
+        CMA operator '''
+        from f2pygen import DeclGen
+        if not operator:
+            raise GenerationError("Internal error: no CMA operator supplied.")
+        if operator.type != "gh_columnwise_operator":
+            raise GenerationError(
+                "Internal error: a CMA operator (gh_columnwise_operator) must "
+                "be supplied but got {0}".format(operator.type))
+        # If a kernel applies a CMA operator then it must only have a
```
Comment (arporter) [inline src/dynamo0p3.py]: That makes two of us. Must be a hangover from something I had during the development process. I've removed the comment.

File: src/dynamo0p3.py [hunk 20]
```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
...
+        # single such operator amongst its arguments
+        if operator.function_space_to.orig_name == function_space.orig_name:
```
Comment (arporter) [inline src/dynamo0p3.py]: I've added a new comment to clarify.

File: src/dynamo0p3.py [hunk 21]
```diff
@@ -3597,11 +4044,20 @@ def get_arg_on_space(self, func_space):
                                  "is no field or operator with function space "
                                  "{0}".format(func_space.mangled_name))
 
-    @property
-    def has_operator(self):
-        ''' Returns true if at least one of the arguments is an operator. '''
+    def has_operator(self, op_type=None):
+        ''' Returns true if at least one of the arguments is an operator
```
Comment (arporter) [inline src/dynamo0p3.py]: doc-string corrected.

File: src/dynamo0p3_builtins.py [hunk 22]
```diff
@@ -150,6 +150,11 @@ def qr_required(self):
     def gen_code(self, parent):
         raise NotImplementedError("DynBuiltIn.gen_code must be overridden")
 
+    def cma_operation(self):
+        ''' Built-ins do not perform operations with Column-Matrix-Assembly
+        operators '''
+        return ""
```
Comment (arporter) [inline src/dynamo0p3_builtins.py]: Good point. Fixed (return None).

File: src/tests/test_files/dynamo0p3/20.0_cma_assembly.f90 [hunk 23]
```diff
@@ -0,0 +1,54 @@
...
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. Ford and A. R. Porter, STFC Daresbury Lab
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
Comment (arporter) [inline src/tests/test_files/dynamo0p3/20.0_cma_assembly.f90]: No! Fixed.

File: src/tests/test_files/dynamo0p3/20.1.1_cma_apply_same_spaces.f90 [hunk 24]
```diff
@@ -0,0 +1,55 @@
...
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. Ford and A. R. Porter, STFC Daresbury Lab
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
Comment (arporter) [inline src/tests/test_files/dynamo0p3/20.1.1_cma_apply_same_spaces.f90]: No :-) Fixed.

File: src/tests/test_files/dynamo0p3/20.1_cma_apply.f90 [hunk 25]
```diff
@@ -0,0 +1,54 @@
...
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. Ford and A. R. Porter, STFC Daresbury Lab
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
Comment (arporter) [inline src/tests/test_files/dynamo0p3/20.1_cma_apply.f90]: Nope.

File: src/tests/test_files/dynamo0p3/20.3_cma_assembly_field.f90 [hunk 26]
```diff
@@ -0,0 +1,53 @@
...
+! CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+! LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+! ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+! POSSIBILITY OF SUCH DAMAGE.
+! -----------------------------------------------------------------------------
+! Author R. Ford and A. R. Porter, STFC Daresbury Lab
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
Comment (arporter) [inline src/tests/test_files/dynamo0p3/20.3_cma_assembly_field.f90]: Fixed.

File: src/tests/test_files/dynamo0p3/columnwise_op_app_same_fs_kernel_mod.F90 [hunk 27]
```diff
@@ -0,0 +1,129 @@
...
+  !> @param [in] map2 dofmap for the from-space
+  !> @param [in] nrow number of rows in the banded matrix
+  !> @param [in] ncol number of columns in the banded matrix
+  !> @param [in] bandwidth bandwidth of the banded matrix
+  !> @param [in] alpha banded matrix parameter \f$\alpha\f$
+  !> @param [in] beta banded matrix parameter \f$\beta\f$
+  !> @param [in] gamma_m banded matrix parameter \f$\gamma_-\f$
+  !> @param [in] gamma_p banded matrix parameter \f$\gamma_+\f$
+  !> @param [in] indirection_dofmap_to indirection map for to-space
+  !> @param [in] indirection_dofmap_from indirection map for from-space
```
Comment (arporter) [inline src/tests/test_files/dynamo0p3/columnwise_op_app_same_fs_kernel_mod.F90]: I've corrected the list. I like having them there as it explains what the various arguments are.

File: src/dynamo0p3.py (hunk 15 above)
Comment (arporter) [inline src/dynamo0p3.py]: Belatedly fixed related test too.

**Comment by arporter** [comment]
@rupertford I think I've addressed everything - back for 2nd review...

File: doc/dynamo0p3.rst (hunk 1 above)
Comment (rupertford) [inline doc/dynamo0p3.rst]: If the dofs are for the whole column then that makes sense but it is unclear from the text. If it could be made more clear that would be good.

File: examples/check_examples (hunk 3 above)
Comment (rupertford) [inline examples/check_examples]: Looks good.

File: examples/dynamo/README (hunk 4 above)
Comment (rupertford) [inline examples/dynamo/README]: Great.

File: src/dynamo0p3.py [hunk 28]
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
...
-                                     "and therefore should only have one CMA "
-                                     "operator argument but found {1}".
-                                     format(self.name, len(cwise_ops)))
-                # We need at least one read-only LMA operator
+            # or performing a matrix-matrix operation...
+            # The kernel must not write to any args other than the CMA
+            # operator
+            write_args = psyGen.args_filter(
+                self._arg_descriptors,
+                arg_types=["gh_operator", "gh_field", "gh_real", "gh_integer"],
```
Comment (rupertford) [inline src/dynamo0p3.py]: Change looks good.

File: src/dynamo0p3.py [hunk 29]
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
...
+                arg_accesses=["gh_inc", "gh_write"])
```
Comment (rupertford) [inline src/dynamo0p3.py]: Change looks good.

File: src/dynamo0p3.py [hunk 30]
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
...
+            if write_args:
+                raise ParseError(
+                    "Kernel {0} writes to a column-wise operator but "
+                    "also writes to {1} argument(s). This is not "
+                    "allowed.".format(self.name,
+                                      [arg.type for arg in write_args]))
+            if len(cwise_ops) == 1:
+
+                # If this is an assembly kernel then we need at least one
```
Comment (rupertford) [inline src/dynamo0p3.py]: Done. Good.

File: src/dynamo0p3.py (hunk 8 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great. Looks better.

File: src/dynamo0p3.py [hunk 31]
```diff
@@ -1188,6 +1200,17 @@ def __init__(self, schedule):
...
+        # We also create a dictionary of column-banded dofmaps. Entries
+        # in this one are a list where the first item is the
+        # corresponding CMA object and the second is whether the map
+        # corresponds to the 'to' or 'from' space.
+        self._unique_cbanded_maps = {}
+        # A dictionary of required CMA indirection dofmaps. Each entry
+        # is a list where the first item correponds to the CMA object
+        # and the second is whether the map corresponds to the 'to' or
+        # 'from' space.
+        self._unique_indirection_maps = {}
```
Comment (rupertford) [inline src/dynamo0p3.py]: Done here too. Great.

File: src/dynamo0p3.py (hunk 9 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great.

File: src/dynamo0p3.py (hunk 10 above)
Comment (rupertford) [inline src/dynamo0p3.py]: OK.

File: src/dynamo0p3.py (hunk 11 above)
Comment (rupertford) [inline src/dynamo0p3.py]: OK, that makes sense.

File: src/dynamo0p3.py (hunk 12 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Your code is now officially better than its surrounding code.

File: src/dynamo0p3.py (hunk 20 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great.

File: doc/dynamo0p3.rst (hunk 1 above)
Comment (arporter) [inline doc/dynamo0p3.rst]: I thought that a dofmap is only ever for a single cell since we have contiguous data-layout in the vertical. i.e. kernels do `map(df)+k` to get the appropriate array index for dof `df` on level `k`. We don't pass in `nlevels` because we the CMA operator acts on whole columns. I've added a line to clarify.

File: src/dynamo0p3.py (hunk 13 above)
Comment (arporter) [inline src/dynamo0p3.py]: Oops, don't know what I did there. I've added a comment in the relevant location.

**Comment by arporter** [comment]
@rupertford I think I've addressed the remaining issues (it's hard to follow - maybe we need to review how we use pull-requests?):

1. I suddenly understood what you meant and have added a comment at 1844 in dynamo0p3.py
2. I've added a line in dynamo0p3.rst to clarify why nlevels is not required when applying a CMA operator
3. I've added a comment saying we choose to pass nrow if ncol==nrow (3471 of dynamo0p3.py).

File: doc/dynamo0p3.rst (hunk 1 above)
Comment (rupertford) [inline doc/dynamo0p3.rst]: I'm really sorry but I still don't understand the text. It says the field passed in is 1D. OK, so we treat the whole column as 1D as it is CMA cf. 2D in "standard" kernels. Fine. It then says the extent of this 1D array is the number of degrees of freedom. Well the degrees of freedom are for a single cell so according to the text the extent of the field passed in is the size of a single cell, not a whole column. Shouldn't the extent be the degrees of freedom for a column?

File: src/dynamo0p3.py (hunk 13 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great.

File: src/dynamo0p3.py (hunk 18 above)
Comment (rupertford) [inline src/dynamo0p3.py]: Great.

File: doc/dynamo0p3.rst (hunk 1 above)
Comment (arporter) [inline doc/dynamo0p3.rst]: Fields are always passed as 1D arrays with dimension `undf` - the unique number of dofs for the function space (See rule 3.2 in the Subroutine section: http://psyclone.readthedocs.io/en/stable/dynamo0p3.html#subroutine). This is not the same as 'ndf' which is the number of dofs *per cell*. I imagine that `undf` is the number of dofs in the local partition and we use the dofmap to pull out the ones relating to the current column/cell. I've updated the doc a little bit to emphasise that ndf is per-cell.

**Comment by arporter** [comment]
@rupertford  - I think I'm done. Realised I needed to correct the copyright notice in some of the columnwise test kernels so I've done that.

File: doc/dynamo0p3.rst (hunk 1 above)
Comment (rupertford) [inline doc/dynamo0p3.rst]: You are quite right. I was getting confused between undf and ndf.
